File: omake_0_9_7_pre7/unix/lm_dll.ml
(*
 * Dynamic link libraries.
 *
 * ----------------------------------------------------------------
 *
 * @begin[license]
 * Copyright (C) 2005 Mojave Group, Caltech
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * Author: Jason Hickey
 * @email{jyh@cs.caltech.edu}
 * @end[license]
 *)
open Lm_string_set
open Lm_symbol
open Lm_printf

(*
 * Flags when a library is opened.
 *)
type open_flag =
   RTLD_LAZY
 | RTLD_NOW
 | RTLD_GLOBAL

type open_flags = open_flag list

(*
 * The info loaded from the DLL.
 *)

(*
 * An object is represented as an array of
 * named values.
 *
 * Runtime object do not include name or type information.
 * They are just an array where the first entry is the name
 * of the object.
 *
 * The type names use C notation, where all typedefs have
 * been unfolded.
 *
 * The values associated with each of the following types
 * are in the normal format.
 *     void         : the value is ()
 *     int          : the value is an integer
 *     char         : same as int
 *     char *       : the value is a string
 *     float        : the value is a float
 *     double       : same as float
 *     object       : the value is an array of values,
 *         where the first element is the name of
 *         the object (so you can get the type information).
 *
 * Values in any other type, like the following, are C pointers.
 * You should treat these values as abstract, and you should carry
 * aroun
 *     void *
 *     struct foo *
 *     void (*)(int, float)
 *)
 *)
type dll_object_field_type =
   { object_field_name : symbol;
     object_field_type : symbol
   }

type dll_object_type =
   { object_name   : symbol;
     object_fields : dll_object_field_type array
   }

(*
 * Enumerations associate integers with values.
 *)
type dll_enum_field =
   { enum_field_name  : symbol;
     enum_field_value : int
   }

type dll_enum =
   { enum_name   : symbol;
     enum_fields : dll_enum_field array
   }

(*
 * A function has a name, and type information for
 * the arguments and the result.
 *
 * The value itself is abstract.
 *)
type 'a dll_value =
   { value_name        : symbol;
     value_fun         : 'a;
     value_arg_types   : symbol array;
     value_result_type : symbol
   }

(*
 * All the info from the DLL.
 *)
type 'a dll =
   { dll_objects : dll_object_type array;
     dll_enums   : dll_enum array;
     dll_values  : 'a dll_value array;
     dll_globals : Obj.t
   }

(*
 * Values to pass to the C functions.
 *)
type dll_val =
   ValUnit
 | ValBool of bool
 | ValInt of int
 | ValFloat of float
 | ValString of string
 | ValObject of symbol * Obj.t array
 | ValAbstract of symbol * Obj.t

(*
 * The raw info loaded from static library.
 *)
type dll_export

(*
 * Types for C values.
 *)
type t_void = unit
type t_char = char
type t_int  = int
type t_float = float
type t_double = float
type t___builtin_va_list

type 'a dll_pointer
type ('args, 'res) dll_function
type dll_object

external p_coerce : 'a dll_pointer -> 'b dll_pointer = "%identity"

(*
 * For creating tuples from singleton values.
 *)
type 'a dll_singleton_tuple =
   SingletonTuple of 'a

(*
 * The DLL interface.
 *)
module type DllSig =
sig
   type t

   type dll_fun
   type dll_sym

   (*
    * Open the DLL.
    *)
   val opendll       : string -> open_flags -> t option

   (*
    * Open the static info.
    *)
   val open_static   : dll_export -> t

   (*
    * Get the value of an enum.
    * This is mainly for use by generated code.
    *)
   val get_enum      : t -> int -> int -> int

   (*
    * Get the global variables.
    *)
   val get_globals   : t -> Obj.t

   (*
    * Get the info of all the values in the DLL.
    *)
   val info           : t -> dll_fun dll

   (*
    * Use this function if you want to handle callbacks.
    * Eventually, it would be good to make the return value general.
    *)
   val set_callback_handler : t -> (Obj.t -> int) -> unit

   (*
    * Apply a function.  This should be reasonably safe, as long
    * as the DLL has reported its type information correctly.
    *)
   val apply : dll_fun -> dll_val array -> dll_val

   (* This is raised if an argument can't be coerced to the right type *)
   exception TypeError of symbol * dll_val

   (* This is raised if the number of arguments doesn't match *)
   exception ArityMismatch of dll_fun * int * int

   (*
    * !!!WARNING!!!
    *
    * The following function bypasses the type checking
    * and calls the DLL function directly.  This is for clients
    * that want to do the typechecking themselves.  The performance
    * is somewhat better.
    *)
   val unsafe_get     : dll_fun -> dll_sym
   val unsafe_apply   : dll_sym -> Obj.t array -> Obj.t

   (*
    * We also need NULL pointers directly.
    *)
   val null           : Obj.t
   val is_null        : 'a -> bool

   val p_NULL         : 'a dll_pointer

   val pointer_of_int : int -> Obj.t
   val int_of_pointer : Obj.t -> int
end;;

module Dll : DllSig =
struct
   (* The abstract type of DLL values *)
   type dll_sym

   (* The function to set the callback handler *)
   type set_handler

   (* Function with type information *)
   type dll_fun = dll_sym dll_value

   (* The DLL info *)
   type t = dll_fun dll * set_handler

   (*
    * The raw information returned by the DLL.
    *)
   type raw_object_field =
      { raw_object_field_name : string;
        raw_object_field_type : string
      }

   type raw_object =
      { raw_object_name : string;
        raw_object_fields : raw_object_field array
      }

   type raw_enum_field =
      { raw_enum_field_name  : string;
        raw_enum_field_value : int
      }

   type raw_enum =
      { raw_enum_name : string;
        raw_enum_fields : raw_enum_field array
      }

   type raw_value =
      { raw_value_name : string;
        raw_value_fun  : dll_sym;
        raw_value_arg_types : raw_object_field array;
        raw_value_result_type : string
      }

   type raw_info =
      Export of raw_object array * raw_enum array * raw_value array * set_handler * Obj.t
    | NoExport

   (* This is raised if an argument can't be coerced to the right type *)
   exception TypeError of symbol * dll_val

   (* This is raised if the number of arguments doesn't match *)
   exception ArityMismatch of dll_fun * int * int

   (************************************************************************
    * Open the DLL.
    *)
   external dlopen  : string -> open_flags -> raw_info = "lm_dlopen";;

   (*
    * Convert the DLL info.
    *)
   let build_object_field info =
      let { raw_object_field_name = name;
            raw_object_field_type = ty
          } = info
      in
         { object_field_name = Lm_symbol.add name;
           object_field_type = Lm_symbol.add ty
         }

   let build_object info =
      let { raw_object_name = name;
            raw_object_fields = fields
          } = info
      in
         { object_name = Lm_symbol.add name;
           object_fields = Array.map build_object_field fields
         }

   let build_enum_field info =
      let { raw_enum_field_name = name;
            raw_enum_field_value = i
          } = info
      in
         { enum_field_name = Lm_symbol.add name;
           enum_field_value = i
         }

   let build_enum info =
      let { raw_enum_name = name;
            raw_enum_fields = fields
          } = info
      in
         { enum_name = Lm_symbol.add name;
           enum_fields = Array.map build_enum_field fields
         }

   let pp_print_string_array buf sl =
      Array.iter (fun s -> fprintf buf " %s," s) sl

   let build_value info =
      let { raw_value_name = name;
            raw_value_fun = sym;
            raw_value_arg_types = arg_types;
            raw_value_result_type = result_type
          } = info
      in
      let info =
         { value_name = Lm_symbol.add name;
           value_fun = sym;
           value_arg_types = Array.map (fun field -> Lm_symbol.add field.raw_object_field_type) arg_types;
           value_result_type = Lm_symbol.add result_type
         }
      in
         { info with value_fun = info }

   let build_dll objects enums values set_handler globals =
      let objects = Array.map build_object objects in
      let enums   = Array.map build_enum enums in
      let values  = Array.map build_value values in
      let info =
         { dll_objects = objects;
           dll_enums   = enums;
           dll_values  = values;
           dll_globals = globals
         }
      in
         info, set_handler

   (*
    * Open and format the DLL.
    *)
   let opendll name flags =
      match dlopen name flags with
         Export (objects, enums, values, set_handler, globals) ->
            Some (build_dll objects enums values set_handler globals)
       | NoExport ->
            None

   let info (x, _) =
      x

   (************************************************************************
    * Static version.
    *)
   external dlopen_static : dll_export -> raw_info = "lm_dlopen_static"

   let open_static info =
      match dlopen_static info with
         Export (objects, enums, values, set_handler, globals) ->
            build_dll objects enums values set_handler globals
       | NoExport ->
            raise (Failure "Lm_dll.open_static")

   let get_enum (info, _) i j =
      info.dll_enums.(i).enum_fields.(j).enum_field_value

   let get_globals (info, _) =
      info.dll_globals

   (************************************************************************
    * Set the callback handler.
    *)
   external set_handler : set_handler -> (Obj.t -> int) -> unit = "lm_set_callback_handler"

   let set_callback_handler (_, info) f =
      set_handler info f

   (************************************************************************
    * Unsafe application.
    *)
   let unsafe_get f =
      f.value_fun

   external unsafe_apply : dll_sym -> Obj.t array -> Obj.t = "lm_dlapply";;

   (*
    * Null pointers.
    *)
   external get_null : unit -> Obj.t = "lm_dlnull";;

   let null = get_null ()

   let p_NULL = Obj.magic null

   let is_null x =
      Obj.repr x == null

   external pointer_of_int : int -> Obj.t = "lm_dlpointer_of_int"
   external int_of_pointer : Obj.t -> int = "lm_dlint_of_pointer"

   (************************************************************************
    * Function calls.
    *)

   (*
    * Type casting.
    *)
   let void_of_value ty v =
      Obj.repr ()

   let int_of_value ty v =
      let v =
         match v with
            ValBool true ->
               1
          | ValUnit
          | ValBool false ->
               0
          | ValInt i ->
               i
          | ValFloat x ->
               int_of_float x
          | ValString s ->
               (try int_of_string s with
                   Failure _ ->
                      raise (TypeError (ty, v)))
          | ValObject _
          | ValAbstract _ ->
               raise (TypeError (ty, v))
      in
         Obj.repr v

   let float_of_value ty v =
      let x =
         match v with
            ValBool true ->
               1.0
          | ValUnit
          | ValBool false ->
               0.0
          | ValInt i ->
               float_of_int i
          | ValFloat x ->
               x
          | ValString s ->
               (try float_of_string s with
                   Failure _ ->
                      raise (TypeError (ty, v)))
          | ValObject _
          | ValAbstract _ ->
               raise (TypeError (ty, v))
      in
         Obj.repr x

   let string_of_value ty v =
      let s =
         match v with
            ValUnit ->
               ""
          | ValBool b ->
               string_of_bool b
          | ValInt i ->
               string_of_int i
          | ValFloat x ->
               string_of_float x
          | ValString s ->
               s
          | ValObject _
          | ValAbstract _ ->
               raise (TypeError (ty, v))
      in
         Obj.repr s

   let c_info =
      ["void",   void_of_value;
       "int",    int_of_value;
       "float",  float_of_value;
       "double", float_of_value;
       "char *", string_of_value]

   let c_table =
      List.fold_left (fun table (s, f) ->
            SymbolTable.add table (Lm_symbol.add s) f) SymbolTable.empty c_info

   let c_of_value ty v =
      try (SymbolTable.find c_table ty) ty v with
         Not_found ->
            match v with
               ValAbstract (ty', v) when Lm_symbol.eq ty' ty ->
                  Obj.repr v
             | _ ->
                  raise (TypeError (ty, v))

   (*
    * Type casting.
    *)
   let value_of_void v =
      ValUnit

   let value_of_int v =
      ValInt (Obj.magic v)

   let value_of_float v =
      ValFloat (Obj.magic v)

   let value_of_string v =
      ValString (Obj.magic v)

   let c_info =
      ["void",   value_of_void;
       "int",    value_of_int;
       "float",  value_of_float;
       "double", value_of_float;
       "char *", value_of_string]

   let c_table =
      List.fold_left (fun table (s, f) ->
            SymbolTable.add table (Lm_symbol.add s) f) SymbolTable.empty c_info

   let value_of_c ty v =
      try (SymbolTable.find c_table ty) v with
         Not_found ->
            ValAbstract (ty, v)

   (*
    * Now the application.
    *)
   let apply f args =
      let { value_fun = sym;
            value_arg_types = arg_types;
            value_result_type = result_type
          } = f
      in
      let len1 = Array.length arg_types in
      let len2 = Array.length args in
      let () =
         if len1 <> len2 then
            raise (ArityMismatch (f, len1, len2))
      in
      let args = Array.mapi (fun i v -> c_of_value (Array.unsafe_get arg_types i) v) args in
      let x = unsafe_apply sym args in
         value_of_c result_type x
end;;

(*!
 * @docoff
 *
 * -*-
 * Local Variables:
 * Caml-master: "compile"
 * End:
 * -*-
 *)

File: cutil/lm_dll.c
/*
 * Dynamic linking.  We provide a minimal amount of type checking.
 *
 * ----------------------------------------------------------------
 *
 * @begin[license]
 * Copyright (C) 2005 Mojave group, Caltech
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation,
 * version 2.1 of the License.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 * Additional permission is given to link this library with the
 * OpenSSL project's "OpenSSL" library, and with the OCaml runtime,
 * and you may distribute the linked executables.  See the file
 * LICENSE.libmojave for more details.
 *
 * Author: Jason Hickey
 * @email{jyh@cs.caltech.edu}
 * @end[license]
 */
#include <stdio.h>
#include <string.h>
#include <assert.h>
#include <caml/mlvalues.h>
#include <caml/alloc.h>
#include <caml/memory.h>
#include <caml/custom.h>
#include <caml/fail.h>
#include <caml/callback.h>
#include "lm_dll.h"

#ifdef WIN32

/*
 * Open a dynamic link library.
 */
value lm_dlopen(value v_filename, value v_flags)
{
    CAMLparam2(v_filename, v_flags);

    failwith("lm_dlopen");

    /* Return it */
    CAMLreturn(Val_unit);
}

/*
 * Open static info.
 */
value lm_dlopen_static(DllExport *values)
{
    CAMLparam0();

    failwith("lm_dlopen_static");

    CAMLreturn(Val_unit);
}

/*
 * Set the callback handler.
 */
typedef void (*SetHandler)(value);
typedef value (*FunPointer)(value);

value lm_set_callback_handler(value v_set, value v_handler)
{
    SetHandler set = (SetHandler) v_set;
    set(v_handler);
    return Val_unit;
}

/*
 * Apply the function.
 */
value lm_dlapply(value v_sym, value v_args)
{
    CAMLparam2(v_sym, v_args);
    failwith("lm_dlapply");
    CAMLreturn(Val_unit);
}

/*
 * NULL pointers.
 */
value lm_dlnull(value v_arg)
{
    CAMLparam1(v_arg);
    CAMLreturn(Val_unit);
}

value lm_dlpointer_of_int(value v_arg)
{
    CAMLparam1(v_arg);
    failwith("lm_dlpointer_of_int");
    CAMLreturn(Val_unit);
}

value lm_dlint_of_pointer(value v_arg)
{
    CAMLparam1(v_arg);
    failwith("lm_dlint_of_pointer");
    CAMLreturn(Val_unit);
}

#else /* !WIN32 */

#include <dlfcn.h>

#define CAML_RTLD_LAZY          0
#define CAML_RTLD_NOW           1
#define CAML_RTLD_GLOBAL        2

/*
 * Hooks.
 */
static DllHooks hooks = {
    caml_alloc,
    caml_alloc_tuple,
    caml_alloc_custom,
    caml_copy_string,
    caml_copy_string_array,
    caml_copy_int32,
    caml_copy_double,
    caml_named_value,
    caml_callback,
    caml_callback2,
    caml_callback_exn,
    caml_callback2_exn,
    caml_register_global_root,
    caml_modify
};

/************************************************************************
 * Pointers are custom blocks.
 */
typedef struct _dll_pointer {
    void *p;
} DllPointer;

static DllPointer dll_null_pointer;

#define DllPointer_val(v)         ((DllPointer *) Data_custom_val(v))
#define DllPointer_pointer_val(v) (DllPointer_val(v)->p)

static int dll_pointer_compare(value v1, value v2)
{
    void *p1 = DllPointer_pointer_val(v1);
    void *p2 = DllPointer_pointer_val(v2);
    int i;

    if(p1 < p2)
        i = -1;
    else if(p1 > p2)
        i = 1;
    else
        i = 0;
    return i;
}

static long dll_pointer_hash(value v)
{
    CAMLparam1(v);
    void *p = DllPointer_pointer_val(v);
    CAMLreturn((long) p);
}

static void dll_pointer_serialize(value v, unsigned long *wsize_32, unsigned long *wsize_64)
{
    *wsize_32 = 4;
    *wsize_64 = 8;
}

static unsigned long dll_pointer_deserialize(void *dst)
{
    *(DllPointer *)dst = dll_null_pointer;
    return sizeof(DllPointer);
}

static struct custom_operations dll_pointer_ops = {
    "dll_pointer",
    custom_finalize_default,
    dll_pointer_compare,
    dll_pointer_hash,
    dll_pointer_serialize,
    dll_pointer_deserialize
};

static inline value dll_unmarshal_pointer(void *p)
{
    value v = alloc_custom(&dll_pointer_ops, sizeof(DllPointer), 0, 1);
    DllPointer_pointer_val(v) = p;
    return v;
}

static inline void *dll_marshal_pointer(value v)
{
    return DllPointer_pointer_val(v);
}

/************************************************************************
 * Allocate of C data.
 */

/*
 * A single field in an object.
 */
static value alloc_field(value v_strings, DllField *field)
{
    CAMLparam1(v_strings);
    CAMLlocal1(v_tuple);

    v_tuple = alloc_tuple(2);
    Store_field(v_tuple, 0, Field(v_strings, field->name));
    Store_field(v_tuple, 1, Field(v_strings, field->type));
    CAMLreturn(v_tuple);
}

/*
 * All the fields in an object or argument list.
 */
static value alloc_rttd(value v_strings, DllField *fields)
{
    CAMLparam1(v_strings);
    CAMLlocal1(v_tuple);
    DllField *p;
    int i;

    /* Get the number of entries */
    i = 0;
    for(p = fields; p->name >= 0; p++)
        i++;
    if(i == 0)
        CAMLreturn(Atom(0));

    /* Alloc them */
    v_tuple = alloc_tuple(i);
    for(i = 0, p = fields; p->name >= 0; i++, p++)
        Store_field(v_tuple, i, alloc_field(v_strings, p));

    CAMLreturn(v_tuple);
}

/*
 * All the objects in the library.
 */
static value alloc_fields(value v_strings, DllField **fields)
{
    CAMLparam1(v_strings);
    CAMLlocal1(v_tuple);
    DllField **p;
    int i;

    /* Get the number of entries */
    i = 0;
    for(p = fields; *p; p++)
        i++;
    if(i == 0)
        CAMLreturn(Atom(0));

    /* Alloc them */
    v_tuple = alloc_tuple(i);
    for(i = 0, p = fields; *p; i++, p++)
        Store_field(v_tuple, i, alloc_rttd(v_strings, *p));
    CAMLreturn(v_tuple);
}

/*
 * A specific object.
 */
static value alloc_object(value v_strings, value v_fields, DllObject *obj)
{
    CAMLparam2(v_strings, v_fields);
    CAMLlocal1(v_tuple);

    v_tuple = alloc_tuple(2);
    Store_field(v_tuple, 0, Field(v_strings, obj->name));
    Store_field(v_tuple, 1, Field(v_fields, obj->fields));
    CAMLreturn(v_tuple);
}

/*
 * All the objects in the program.
 */
static value alloc_objects(value v_strings, value v_fields, DllObject *objs)
{
    CAMLparam2(v_strings, v_fields);
    CAMLlocal1(v_tuple);
    DllObject *p;
    int i;

    /* Get the number of entries */
    i = 0;
    for(p = objs; p->name >= 0; p++)
        i++;
    if(i == 0)
        CAMLreturn(Atom(0));

    /* Alloc them */
    v_tuple = alloc_tuple(i);
    for(i = 0, p = objs; p->name >= 0; i++, p++)
        Store_field(v_tuple, i, alloc_object(v_strings, v_fields, p));

    CAMLreturn(v_tuple);
}

/*
 * Enumerations.
 */
static value alloc_enum_field(value v_strings, DllEnumField *field)
{
    CAMLparam1(v_strings);
    CAMLlocal1(v_tuple);

    v_tuple = alloc_tuple(2);
    Store_field(v_tuple, 0, Field(v_strings, field->name));
    Store_field(v_tuple, 1, Val_int(field->value));
    CAMLreturn(v_tuple);
}

static value alloc_enum_fields(value v_strings, DllEnumField *fields)
{
    CAMLparam1(v_strings);
    CAMLlocal1(v_tuple);
    DllEnumField *p;
    int i;

    /* Get the number of entries */
    i = 0;
    for(p = fields; p->name >= 0; p++)
        i++;
    if(i == 0)
        CAMLreturn(Atom(0));

    /* Alloc them */
    v_tuple = alloc_tuple(i);
    for(i = 0, p = fields; p->name >= 0; i++, p++)
        Store_field(v_tuple, i, alloc_enum_field(v_strings, p));

    CAMLreturn(v_tuple);
}

static value alloc_enum_info(value v_strings, DllEnumInfo *info)
{
    CAMLparam1(v_strings);
    CAMLlocal1(v_tuple);

    v_tuple = alloc_tuple(2);
    Store_field(v_tuple, 0, Field(v_strings, info->name));
    Store_field(v_tuple, 1, alloc_enum_fields(v_strings, info->fields));
    CAMLreturn(v_tuple);
}

static value alloc_enums(value v_strings, DllEnumInfo *enums)
{
    CAMLparam1(v_strings);
    CAMLlocal1(v_tuple);
    DllEnumInfo *p;
    int i;

    /* Get the number of entries */
    i = 0;
    for(p = enums; p->name >= 0; p++)
        i++;
    if(i == 0)
        CAMLreturn(Atom(0));

    /* Alloc them */
    v_tuple = alloc_tuple(i);
    for(i = 0, p = enums; p->name >= 0; p++, i++)
        Store_field(v_tuple, i, alloc_enum_info(v_strings, p));

    CAMLreturn(v_tuple);
}

/*
 * A single function value.
 */
static value alloc_value(value v_strings, value v_fields, DllValue *val)
{
    CAMLparam2(v_strings, v_fields);
    CAMLlocal1(v_tuple);

    v_tuple = alloc_tuple(4);
    Store_field(v_tuple, 0, Field(v_strings, val->name));
    Store_field(v_tuple, 1, dll_unmarshal_pointer(val->value));
    Store_field(v_tuple, 2, Field(v_fields, val->arg_types));
    Store_field(v_tuple, 3, Field(v_strings, val->result_type));
    CAMLreturn(v_tuple);
}

/*
 * All the values in the program.
 */
static value alloc_values(value v_strings, value v_fields, DllValue *values)
{
    CAMLparam1(v_strings);
    CAMLlocal1(v_tuple);
    DllValue *p;
    int i;

    /* Get the number of entries */
    i = 0;
    for(p = values; p->name >= 0; p++)
        i++;
    if(i == 0)
        CAMLreturn(Atom(0));

    /* Alloc them */
    v_tuple = alloc_tuple(i);
    for(i = 0, p = values; p->name >= 0; i++, p++)
        Store_field(v_tuple, i, alloc_value(v_strings, v_fields, p));

    CAMLreturn(v_tuple);
}

/*
 * Allocate everything.
 */
static value alloc_dll(DllExport *values, value v_globals)
{
    CAMLparam1(v_globals);
    CAMLlocal3(v_strings, v_fields, v_tuple);

    /* Allocate the strings */
    v_strings = copy_string_array((const char **) values->strings);
    v_fields = alloc_fields(v_strings, values->fields);

    /* Allocate the result */
    v_tuple = alloc_tuple(5);
    Store_field(v_tuple, 0, alloc_objects(v_strings, v_fields, values->objects));
    Store_field(v_tuple, 1, alloc_enums(v_strings, values->enums));
    Store_field(v_tuple, 2, alloc_values(v_strings, v_fields, values->values));
    Store_field(v_tuple, 3, (value) values->set_callback_handler);
    Store_field(v_tuple, 4, v_globals);
    CAMLreturn(v_tuple);
}

/************************************************************************
 * DLL functions.
 */

/*
 * Open a dynamic link library.
 */
value lm_dlopen(value v_filename, value v_flags)
{
    CAMLparam2(v_filename, v_flags);
    CAMLlocal1(v_globals);
    DllExport *values;
    int flags;
    void *dll;
    value v;

    /* Parse the flags */
    flags = RTLD_LAZY;
    while(Is_block(v_flags)) {
        value v_flag = Field(v_flags, 0);
        v_flags = Field(v_flags, 1);
        switch(Int_val(v_flag)) {
        case CAML_RTLD_LAZY:
            flags = (flags & ~RTLD_NOW) | RTLD_LAZY;
            break;
        case CAML_RTLD_NOW:
            flags = (flags & ~RTLD_LAZY) | RTLD_NOW;
            break;
        case CAML_RTLD_GLOBAL:
            flags |= RTLD_GLOBAL;
            break;
        default:
            caml_invalid_argument("caml_dlopen");
            break;
        }
    }

    /* Now open the lib */
    dll = dlopen(String_val(v_filename), flags);
    if(dll == 0)
        failwith(dlerror());

    /* Load the value description */
    values = dlsym(dll, "dll_export");
    if(values) {
        if(values->magic != LM_DLL_MAGIC) {
            dlclose(dll);
            failwith(String_val(v_filename));
        }
        v_globals = values->initialize_dll(&hooks);
        v = alloc_dll(values, v_globals);
    }
    else
        v = Val_int(0);

    /* Return it */
    CAMLreturn(v);
}

/*
 * Open static info.
 */
value lm_dlopen_static(DllExport *values)
{
    CAMLparam0();
    CAMLlocal2(v_globals, v);

    if(values->magic == LM_DLL_MAGIC) {
        v_globals = values->initialize_dll(&hooks);
        v = alloc_dll(values, v_globals);
    }
    else
        v = Val_int(0);
    CAMLreturn(v);
}

/*
 * Set the callback handler.
 */
typedef void (*SetHandler)(value);
typedef value (*FunPointer)(value);

value lm_set_callback_handler(value v_set, value v_handler)
{
    SetHandler set = (SetHandler) v_set;
    set(v_handler);
    return Val_unit;
}

/*
 * Apply the function.
 */
value lm_dlapply(value v_sym, value v_args)
{
    CAMLparam2(v_sym, v_args);
    CAMLlocal1(v_result);
    FunPointer f;

    f = (FunPointer) dll_marshal_pointer(v_sym);
    v_result = f(v_args);
    CAMLreturn(v_result);
}

/*
 * NULL pointers.
 */
value lm_dlnull(value v_arg)
{
    return dll_unmarshal_pointer(0);
}

value lm_dlpointer_of_int(value v_arg)
{
    return dll_unmarshal_pointer((void *) Int_val(v_arg));
}

value lm_dlint_of_pointer(value v_arg)
{
    CAMLparam1(v_arg);
    CAMLlocal1(v);

    if(Is_long(v_arg))
        v = v_arg;
    else if(Tag_val(v_arg) == Custom_tag && strcmp(Custom_ops_val(v_arg)->identifier, "dll_pointer") == 0)
        v = Val_int(dll_marshal_pointer(v_arg));
    else
        failwith("int_of_pointer: not a pointer or integer");
    CAMLreturn(v);
}

#endif /* !WIN32 */
